File: OsmAnd-java/src/main/java/net/osmand/gpx/SplitMetric.java
package net.osmand.gpx;

import net.osmand.gpx.GPXUtilities.TrkSegment;
import net.osmand.gpx.GPXUtilities.WptPt;

import java.util.List;

abstract class SplitMetric {

	public abstract double metric(WptPt p1, WptPt p2);


	static class DistanceSplitMetric extends SplitMetric {

		private final float[] calculations = new float[1];

		@Override
		public double metric(WptPt p1, WptPt p2) {
			net.osmand.Location.distanceBetween(p1.lat, p1.lon, p2.lat, p2.lon, calculations);
			return calculations[0];
		}
	}

	static class TimeSplitMetric extends SplitMetric {

		@Override
		public double metric(WptPt p1, WptPt p2) {
			if (p1.time != 0 && p2.time != 0) {
				return (int) Math.abs((p2.time - p1.time) / 1000l);
			}
			return 0;
		}
	}

	static void splitSegment(SplitMetric metric, SplitMetric secondaryMetric, double metricLimit, List<SplitSegment> splitSegments, TrkSegment segment, boolean joinSegments) {
		double currentMetricEnd = metricLimit;
		double secondaryMetricEnd = 0;
		SplitSegment sp = new SplitSegment(segment, 0, 0);
		double total = 0;
		WptPt prev = null;
		for (int k = 0; k < segment.points.size(); k++) {
			WptPt point = segment.points.get(k);
			if (k > 0) {
				double currentSegment = 0;
				if (!(segment.generalSegment && !joinSegments && point.firstPoint)) {
					currentSegment = metric.metric(prev, point);
					secondaryMetricEnd += secondaryMetric.metric(prev, point);
				}
				while (total + currentSegment > currentMetricEnd) {
					double p = currentMetricEnd - total;
					double cf = (p / currentSegment);
					sp.setLastPoint(k - 1, cf);
					sp.metricEnd = currentMetricEnd;
					sp.secondaryMetricEnd = secondaryMetricEnd;
					splitSegments.add(sp);

					sp = new SplitSegment(segment, k - 1, cf);
					currentMetricEnd += metricLimit;
				}
				total += currentSegment;
			}
			prev = point;
		}
		if (segment.points.size() > 0 && !(sp.endPointInd == segment.points.size() - 1 && sp.startCoeff == 1)) {
			sp.metricEnd = total;
			sp.secondaryMetricEnd = secondaryMetricEnd;
			sp.setLastPoint(segment.points.size() - 2, 1);
			splitSegments.add(sp);
		}
	}
}

File: OsmAnd-shared/src/commonMain/kotlin/net/osmand/shared/gpx/SplitMetric.kt
package net.osmand.shared.gpx

import kotlin.math.*
import net.osmand.shared.gpx.primitives.WptPt
import net.osmand.shared.gpx.primitives.TrkSegment
import net.osmand.shared.extensions.toRadians

abstract class SplitMetric {

	abstract fun metric(p1: WptPt, p2: WptPt): Double

	class DistanceSplitMetric : SplitMetric() {
		private val calculations = FloatArray(1)

		override fun metric(p1: WptPt, p2: WptPt): Double {
			// Replace the net.osmand.Location.distanceBetween method with an appropriate implementation
			// Here is a simple haversine formula for demonstration purposes
			val R = 6371e3 // Radius of the earth in meters
			val lat1 = p1.lat.toRadians()
			val lat2 = p2.lat.toRadians()
			val dLat = (p2.lat - p1.lat).toRadians()
			val dLon = (p2.lon - p1.lon).toRadians()

			val a = sin(dLat / 2) * sin(dLat / 2) +
					cos(lat1) * cos(lat2) *
					sin(dLon / 2) * sin(dLon / 2)
			val c = 2 * atan2(sqrt(a), sqrt(1 - a))

			return R * c
		}
	}

	class TimeSplitMetric : SplitMetric() {
		override fun metric(p1: WptPt, p2: WptPt): Double {
			return if (p1.time != 0L && p2.time != 0L) {
				abs((p2.time - p1.time) / 1000L).toDouble()
			} else {
				0.0
			}
		}
	}

	companion object {
		fun splitSegment(
			metric: SplitMetric, secondaryMetric: SplitMetric, metricLimit: Double,
			splitSegments: MutableList<SplitSegment>, segment: TrkSegment, joinSegments: Boolean
		) {
			var currentMetricEnd = metricLimit
			var secondaryMetricEnd = 0.0
			var sp = SplitSegment(segment, 0, 0.0)
			var total = 0.0
			var prev: WptPt? = null

			for (k in segment.points.indices) {
				val point = segment.points[k]
				if (k > 0) {
					var currentSegment = 0.0
					if (!(segment.generalSegment && !joinSegments && point.firstPoint)) {
						prev?.let {
							currentSegment = metric.metric(it, point)
							secondaryMetricEnd += secondaryMetric.metric(it, point)
						}
					}
					while (total + currentSegment > currentMetricEnd) {
						val p = currentMetricEnd - total
						val cf = p / currentSegment
						sp.setLastPoint(k - 1, cf)
						sp.metricEnd = currentMetricEnd
						sp.secondaryMetricEnd = secondaryMetricEnd
						splitSegments.add(sp)

						sp = SplitSegment(segment, k - 1, cf)
						currentMetricEnd += metricLimit
					}
					total += currentSegment
				}
				prev = point
			}
			if (segment.points.isNotEmpty() && !(sp.endPointInd == segment.points.size - 1 && sp.startCoeff == 1.0)) {
				sp.metricEnd = total
				sp.secondaryMetricEnd = secondaryMetricEnd
				sp.setLastPoint(segment.points.size - 2, 1.0)
				splitSegments.add(sp)
			}
		}
	}
}
